### PR TITLE
#11132 Run Post-Commit Python Tests against wheels 

### PR DESCRIPTION
**Pull request by dimitri-tenstorrent**
The purpose of this PR is to introduce running of Post Commit Python tests in the isolated Docker environment with the Python environment setup by installing Wheels.

### Ticket

#8082
#11132
#11257

### Problem description

1. We need to make sure the wheels that we publicize to our users have been tested. 
2. We need to test the Release version of Docker image with wheels installed ( this will need the definition of the release image)
3. Also, we need to be moving towards dockerization of CI/CD and this is a small step towards that.

### What's changed

The order of jobs in All Post Commit Workflow was re-arranged to set `build-wheels` job as a prerequisite of jobs containing Python tests. The wheels that are built are installed into the Docker environment and these resulting Docker images are pushed to Docker registry. 
The Python Tests `fast-dispatch-unit-tests`, `ttnn-unit-tests` and `model-unit-tests` are now executed inside of the Docker container with wheel.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: .github/workflows/models-post-commit.yaml [hunk 1]
```diff
@@ -57,19 +57,23 @@ jobs:
...
-        timeout-minutes: ${{ inputs.timeout }}
-        run: |
-          source ${{ github.workspace }}/python_env/bin/activate
-          cd $TT_METAL_HOME
-          export PYTHONPATH=$TT_METAL_HOME
-          ${{ matrix.test-group.cmd }}
+          docker_username: ${{ github.actor }}
+          docker_password: ${{ secrets.GITHUB_TOKEN }}
+          docker_image_arch: ${{ inputs.arch }}
+          docker_opts: |
```
Comment (dimitri-tenstorrent) [inline .github/workflows/models-post-commit.yaml:67]: Move some of the repeating options into the Docker Run command.

File: .github/workflows/build-docker-with-wheels-artifact.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            echo "IMAGE_TAG=dev-${GITHUB_ACTOR}.${{ matrix.arch }}.${{ github.sha }}" >> $GITHUB_ENV
+          fi
+      - name: Build Docker image
+        run: |
+          ARCH_NAME=${{ matrix.arch}}
+          # Do we not have to unzip? 
+          # unzip eager-dist-${{ matrix.os }}-${{ matrix.arch }}.zip
+          WHEEL_FILENAME=$(ls -1 *.whl)
+          echo $WHEEL_FILENAME
+          docker build -f dockerfile/tt-metalium.Dockerfile -t ${{ env.TT_METAL_DOCKER_IMAGE_TAG }}-${{ matrix.arch }} . --build-arg="WHEEL_FILENAME=${WHEEL_FILENAME}" --build-arg="IMAGE_TAG=dev-${GITHUB_ACTOR}"
```
Comment (dimitri-tenstorrent) [inline .github/workflows/build-docker-with-wheels-artifact.yaml:79]: Create a separate workflow to create a release image and run tests on it.

File: dockerfile/ubuntu-20.04-amd64.Dockerfile [hunk 3]
```diff
@@ -33,15 +33,16 @@ ENV PYTHON_ENV_DIR=${TT_METAL_INFRA_DIR}/tt-metal/python_env
 # ENV PATH="$PYTHON_ENV_DIR/bin:$PATH"
 
 # Copy requirements from tt-metal folders with requirements.txt docs
-COPY /docs/requirements-docs.txt ${TT_METAL_INFRA_DIR}/tt-metal/docs/.
+COPY /docs/requirements-docs.txt ${TT_METAL_INFRA_DIR}/tt-metal/docs/requirements-docs.txt
```
Comment (dimitri-tenstorrent) [inline dockerfile/ubuntu-20.04-amd64.Dockerfile:36]: Is everything here necessary for running of Python tests? This is related to the effort of creation of Release image - create a separate ticket to track that work.

File: .github/workflows/eager-package-main.yaml [hunk 4]
```diff
@@ -17,30 +17,30 @@ jobs:
...
-          name: eager-dist-${{ matrix.os }}-${{ matrix.arch }}
-      - name: Set up end-to-end tests environment
-        run: ./tests/scripts/set_up_end_to_end_tests_env.sh
-      - name: Activate env and run release tests - host
-        shell: bash
-        run: |
-          source tests/end_to_end_tests/env/bin/activate
-          cd tests/end_to_end_tests
-          pytest -c conftest.py . -m eager_host_side
+  # test-eager-package-host:
```
Comment (dimitri-tenstorrent) [inline .github/workflows/eager-package-main.yaml:20]: Fix this test - it has some issues with dependencies being all haywire.

File: tests/end_to_end_tests/requirements.txt [hunk 5]
```diff
@@ -1 +1,2 @@
 -r ../../tt_metal/python_env/requirements-dev.txt
+wheel
```
Comment (dimitri-tenstorrent) [inline tests/end_to_end_tests/requirements.txt:2]: Why is this necessary? It seems we are not installing the wheel package but how does it work now?

File: tests/scripts/set_up_end_to_end_tests_env.sh [hunk 6]
```diff
@@ -17,6 +17,7 @@ set_up_end_to_end_tests_env() {
   source env/bin/activate
 
   python -m pip config set global.extra-index-url https://download.pytorch.org/whl/cpu
+  python -m pip install setuptools wheel
```
Comment (dimitri-tenstorrent) [inline tests/scripts/set_up_end_to_end_tests_env.sh:20]: Related to https://github.com/tenstorrent/tt-metal/pull/11870/files#r1731731907

File: .github/workflows/build-docker-with-wheels-artifact.yaml (hunk 2 above)
Comment (dimitri-tenstorrent) [reply .github/workflows/build-docker-with-wheels-artifact.yaml:79]: Create an issue

File: dockerfile/ubuntu-20.04-amd64.Dockerfile (hunk 3 above)
Comment (dimitri-tenstorrent) [reply dockerfile/ubuntu-20.04-amd64.Dockerfile:36]: Create an issue

File: .github/workflows/all-post-commit-workflows.yaml [hunk 7]
```diff
@@ -11,11 +11,20 @@ permissions:
...
 
 jobs:
   static-checks:
     uses: ./.github/workflows/all-static-checks.yaml
     secrets: inherit
+  build-wheels:
+    needs: build-artifact
+    uses: ./.github/workflows/eager-package-main.yaml
+    secrets: inherit
+  build-docker-with-wheels-artifact:
```
Comment (dimitri-tenstorrent) [inline .github/workflows/all-post-commit-workflows.yaml:24]: Run an experiment to test performance of just installing the wheel before test execution. If acceptable, remove this step.

**Comment by dimitri-tenstorrent** [comment]
TODO: Squash Commits

File: .github/workflows/build-docker-artifact.yaml [hunk 8]
```diff
@@ -1,18 +1,26 @@
 name: "Build tt-metal docker artifact"
 
 on:
-  workflow_call:
   workflow_dispatch:
-
+  push:
+    branches:
+      - dimitri/8082-run-python-tests-on-wheel # Needs to be changed before the merge
```
Comment (dimitri-tenstorrent) [inline .github/workflows/build-docker-artifact.yaml:7]: Change to main upon merge.

File: .github/workflows/all-post-commit-workflows.yaml (hunk 7 above)
Comment (dimitri-tenstorrent) [reply .github/workflows/all-post-commit-workflows.yaml:24]: Waiting ....

File: .github/workflows/build-docker-with-wheels-artifact.yaml (hunk 2 above)
Comment (dimitri-tenstorrent) [reply .github/workflows/build-docker-with-wheels-artifact.yaml:77]: https://github.com/tenstorrent/tt-metal/issues/11973

File: dockerfile/ubuntu-20.04-amd64.Dockerfile (hunk 3 above)
Comment (dimitri-tenstorrent) [reply dockerfile/ubuntu-20.04-amd64.Dockerfile:36]: https://github.com/tenstorrent/tt-metal/issues/11974

**Comment by dimitri-tenstorrent** [comment]
> TODO: Squash Commits

This is not needed with the merge commits.

File: tests/tt_eager/python_api_testing/conv/pytorch_conv_tb.py [hunk 9]
```diff
@@ -17,6 +17,41 @@ class TestLevel(Enum):
     OP_FULL_COMPUTE = 4
 
 
+# Moved from conv_sweep_params.yaml
+# Remove when the issue is fixed https://github.com/tenstorrent/tt-metal/issues/11257
```
Comment (dimitri-tenstorrent) [inline tests/tt_eager/python_api_testing/conv/pytorch_conv_tb.py:21]: This is needed due to #11257

File: .github/workflows/build-docker-artifact.yaml [hunk 10]
```diff
@@ -1,18 +1,26 @@
...
+    paths:
+      - dockerfile/ubuntu-20.04-amd64.Dockerfile
```
Comment (TT-billteng) [inline .github/workflows/build-docker-artifact.yaml:9]: what changes would be needed to do this for 22.04 also?

File: .github/workflows/build-docker-with-wheels-artifact.yaml [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        required: false
+        type: boolean
+        default: false
+        description: "Build with tracy enabled"
+jobs:
+  build-docker-images:
+    strategy:
+      matrix:
+        arch: ${{ fromJson(inputs.arch || '["grayskull", "wormhole_b0", "blackhole"]') }}
+        os: [ ubuntu-20.04 ]
```
Comment (TT-billteng) [inline .github/workflows/build-docker-with-wheels-artifact.yaml:43]: should look into generalizing for 22.04 / 24.04 support in a follow-on PR

File: scripts/docker/requirements-20.04.txt [hunk 12]
```diff
@@ -19,3 +19,7 @@ python3-dev=3.8.2-0ubuntu2
 python3.8-venv
 cargo
 ninja-build
+libyaml-cpp-dev=0.6.2-4ubuntu1
```
Comment (tt-rkim) [inline scripts/docker/requirements-20.04.txt:22]: Hmmm do we need all this? Are you sure?

I think we can delete boost and yaml-cpp for example.

File: .github/workflows/ttnn-post-commit.yaml [hunk 13]
```diff
@@ -73,23 +69,19 @@ jobs:
...
         run: |
           echo "TTNN_CONFIG_OVERRIDES={\"enable_fast_runtime_mode\": false}" >> $GITHUB_ENV
       - name: ${{ matrix.test-group.name }} tests
-        timeout-minutes: ${{ inputs.timeout }}
-        run: |
-          source ${{ github.workspace }}/python_env/bin/activate
-          cd $TT_METAL_HOME
-          export PYTHONPATH=$TT_METAL_HOME
-          ${{ matrix.test-group.cmd }}
+        timeout-minutes: 120
```
Comment (tt-rkim) [inline .github/workflows/ttnn-post-commit.yaml:81]: Seems like quite the timeout!

File: .github/actions/docker-run/action.yml [hunk 14]
```diff
@@ -59,11 +65,13 @@ runs:
...
           --net=host
           ${{ inputs.docker_opts }}
+          -e LOGURU_LEVEL=${{ env.LOGURU_LEVEL }}
+          -e PYTHONPATH=/app/mtl
+          -e ARCH_NAME=${{ inputs.docker_image_arch }}
+          ${{ inputs.device }}
         run: |
+          cd /app/mtl
+          rm -rf tt_metal tt_eager
           ${{ inputs.run_args }}
```
Comment (tt-rkim) [inline .github/actions/docker-run/action.yml]: POSIX newline!

File: .github/actions/docker-run/action.yml [hunk 15]
```diff
@@ -59,11 +65,13 @@ runs:
...
-          -v /home/ubuntu/.cache/torch:/home/ubuntu/.cache/torch
           --net=host
           ${{ inputs.docker_opts }}
+          -e LOGURU_LEVEL=${{ env.LOGURU_LEVEL }}
+          -e PYTHONPATH=/app/mtl
+          -e ARCH_NAME=${{ inputs.docker_image_arch }}
+          ${{ inputs.device }}
         run: |
+          cd /app/mtl
+          rm -rf tt_metal tt_eager
```
Comment (tt-rkim) [inline .github/actions/docker-run/action.yml:72]: Why are we deleting these?

File: tests/scripts/set_up_end_to_end_tests_env.sh (hunk 6 above)
Comment (tt-rkim) [reply tests/scripts/set_up_end_to_end_tests_env.sh:20]: Related to my question above. What did you see that led you to add this?

File: tests/end_to_end_tests/requirements.txt (hunk 5 above)
Comment (tt-rkim) [reply tests/end_to_end_tests/requirements.txt:2]: Haha, why is this necessary?

What did you see in your testing that forced this? The wheel should come from `eager-package`, which doesn't seem to be run in a docker container still?

File: .github/workflows/cpp-post-commit.yaml [hunk 16]
```diff
@@ -44,6 +44,8 @@ jobs:
       matrix:
         test-group: [
           {name: C++, cmd: ./tests/scripts/run_cpp_unit_tests.sh},
+          {name: ttnn examples and cpp tests, cmd: ./build/test/ttnn/unit_tests_ttnn && ./tests/scripts/run_ttnn_examples.sh},
```
Comment (tt-rkim) [inline .github/workflows/cpp-post-commit.yaml:47]: Can we move `run_ttnn_examples.sh` back into the ttnn post commit? That's a python test.

File: .github/workflows/build-docker-with-wheels-artifact.yaml (hunk 2 above)
Comment (tt-rkim) [reply .github/workflows/build-docker-with-wheels-artifact.yaml:77]: I would personally keep this file on another branch rather than this PR, and work on it on the side for release images for that issue.

Just to separate concerns.

File: .github/workflows/build-docker-artifact.yaml
Comment (tt-rkim) [inline .github/workflows/build-docker-artifact.yaml:1]: There may be some conflicts with @ttmchiou 's changes here

File: .github/workflows/all-post-commit-workflows.yaml [hunk 17]
```diff
@@ -11,11 +11,16 @@ permissions:
...
   pages: write
   id-token: write
+  packages: write
 
 jobs:
   static-checks:
     uses: ./.github/workflows/all-static-checks.yaml
     secrets: inherit
+  build-wheels:
+    needs: build-artifact
```
Comment (tt-rkim) [inline .github/workflows/all-post-commit-workflows.yaml:21]: Haha great renaming

Wonder if we can eventually merge wheels and binary artifacts jobs? @TT-billteng

File: .github/actions/load-docker-image/action.yml
Comment (tt-rkim) [inline .github/actions/load-docker-image/action.yml:1]: Looks unused. Can we delete?

File: .github/workflows/build-docker-artifact.yaml
Comment (dimitri-tenstorrent) [reply .github/workflows/build-docker-artifact.yaml:1]: Yes, absolutely - I am going to move them to another branch that is about releases!

👍

Comment (dimitri-tenstorrent) [reply .github/workflows/build-docker-artifact.yaml:1]: https://github.com/tenstorrent/tt-metal/tree/dimitri/11973-split-dockerfiles-into-multiple-dockerfiles-for-transparency-and-modularity-base-dev-release

File: tests/end_to_end_tests/requirements.txt (hunk 5 above)
Comment (dimitri-tenstorrent) [reply tests/end_to_end_tests/requirements.txt]: I think these are leftovers from my testing - works now without it.

File: .github/actions/docker-run/action.yml (hunk 15 above)
Comment (dimitri-tenstorrent) [reply .github/actions/docker-run/action.yml:72]: We do not want to mount the cache folders inside of the Docker container and have it be writeable by the user in a different OS. At least that is my opinion, too many issues with permissions. 

I am pretty open about this but feels wrong to me to map the host user's internal folder to the local Docker user.

File: .github/workflows/build-docker-artifact.yaml (hunk 8 above)
Comment (ttmchiou) [reply .github/workflows/build-docker-artifact.yaml:7]: Curious, I think there can be a disconnect if the user is not sure the docker image has finished building yet?

Example:
A user can push a commit up with changes for dockerfile
and immediately run post-commit. This would be that this particular post-commit run can start running before the new docker image is and ready for use?

File: .github/workflows/build-docker-artifact.yaml (hunk 10 above)
Comment (ttmchiou) [reply .github/workflows/build-docker-artifact.yaml:9]: does this support wildcards?
we can do something like -dockerfile/*.Dockerfile or something of that sort

File: .github/actions/docker-run/action.yml (hunk 15 above)
Comment (TT-billteng) [reply .github/actions/docker-run/action.yml:72]: What is $ROOT in this case? Wouldn't this clear out a user's directory if it was mapped into the docker?

Comment (TT-billteng) [reply .github/actions/docker-run/action.yml:72]: If we're using this action with the dev / kitchen sink docker, I'd think this would get mapped into the docker from the outside?

File: .github/workflows/all-post-commit-workflows.yaml (hunk 17 above)
Comment (TT-billteng) [reply .github/workflows/all-post-commit-workflows.yaml:21]: you mean build artifact workflow would create both?

File: scripts/docker/requirements-20.04.txt [hunk 18]
```diff
@@ -19,3 +19,5 @@ python3-dev=3.8.2-0ubuntu2
 python3.8-venv
 cargo
 ninja-build
+patchelf
```
Comment (ttmchiou) [inline scripts/docker/requirements-20.04.txt:22]: I think we are removing this in a separate PR?
https://github.com/tenstorrent/tt-metal/issues/11782

File: .github/workflows/all-post-commit-workflows.yaml (hunk 17 above)
Comment (dimitri-tenstorrent) [reply .github/workflows/all-post-commit-workflows.yaml:21]: I do not see why not. But not this PR :)

Comment (TT-billteng) [reply .github/workflows/all-post-commit-workflows.yaml:21]: true that can work :D

Comment (TT-billteng) [reply .github/workflows/all-post-commit-workflows.yaml:21]: only disadvantage would be that C++ tests could only start after wheel is built, which it wouldn't need

File: tests/scripts/run_python_model_tests.sh [hunk 19]
```diff
@@ -20,40 +15,40 @@ fi
 
 if [ "$ARCH_NAME" != "wormhole_b0" ]; then
     # Tests for tensors in L1
-    pytest $TT_METAL_HOME/models/experimental/bert_large_performant/unit_tests/test_bert_large*matmul* -k in0_L1-in1_L1-bias_L1-out_L1
-    pytest $TT_METAL_HOME/models/experimental/bert_large_performant/unit_tests/test_bert_large*bmm* -k in0_L1-in1_L1-out_L1
+    pytest models/experimental/bert_large_performant/unit_tests/test_bert_large*matmul* -k in0_L1-in1_L1-bias_L1-out_L1
```
Comment (TT-billteng) [inline tests/scripts/run_python_model_tests.sh:18]: `$TT_METAL_HOME` isn't needed even if you run outside of docker now?

**Review by TT-billteng — APPROVED** [review]
Great work!!!!

File: dockerfile/tt-metalium.Dockerfile [hunk 20]
```diff
@@ -0,0 +1,25 @@
+ARG IMAGE_TAG=latest
```
Comment (dimitri-tenstorrent) [inline dockerfile/tt-metalium.Dockerfile:1]: Technically this file also belongs in the Docker Release branch

File: .github/actions/docker-run/action.yml (hunk 15 above)
Comment (dimitri-tenstorrent) [reply .github/actions/docker-run/action.yml:71]: > Why are we deleting these?

Because they are conflicting with installed Python packages. Let's say we are importing `tt_metal` is that done from site-packages or from the current folder?

File: scripts/docker/requirements-20.04.txt (hunk 18 above)
Comment (tt-rkim) [reply scripts/docker/requirements-20.04.txt:22]: Yes, we are https://github.com/tenstorrent/tt-metal/pull/11884

File: dockerfile/ubuntu-20.04-amd64.Dockerfile (hunk 3 above)
Comment (tt-rkim) [reply dockerfile/ubuntu-20.04-amd64.Dockerfile:36]: Depends on if you run the docs pipeline in docker as well , in which case yes, but we don't have to.

File: .github/workflows/build-docker-artifact.yaml (hunk 10 above)
Comment (tt-rkim) [reply .github/workflows/build-docker-artifact.yaml:9]: Pretty sure it does, fam

File: .github/actions/docker-run/action.yml (hunk 15 above)
Comment (tt-rkim) [reply .github/actions/docker-run/action.yml:71]: Sounds good, should remove `ttnn` and as well then.

**Review by tt-rkim — APPROVED** [review]
looks good man, some cleanup things:

- could move the username and password in the action to default to the actor and `GITHUB_TOKEN` so we don't have to copy paste it everywhere
